### PR TITLE
[SPARK-35766][SQL][TESTS] Break down CastSuite/AnsiCastSuite into multiple files

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Currently, the file CastSuite.scala becomes big: 2000 lines, 2 base classes, 4 test suites.
In my previous work of Timestamp without time zone, I planned to put new test cases in CastSuiteBase, but they were accidentally added in AnsiCastSuiteBase.

This PR is to break the file down into 3 files. It also moves the test cases about timestamp without time zone to the right base class.

### Why are the changes needed?

Make development and review easier.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AnsiCastSuiteBase.scala [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+      checkExceptionInExpression[ArithmeticException](
+        cast(value, dt), "overflow")
+      checkExceptionInExpression[ArithmeticException](
+        cast((value * Decimal(1.1)).toFloat, dt), "overflow")
+      checkExceptionInExpression[ArithmeticException](
+        cast((value * Decimal(1.1)).toDouble, dt), "overflow")
+    }
+  }
+
+  test("ANSI mode: Throw exception on casting out-of-range value to byte type") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AnsiCastSuiteBase.scala:54]: we can override the `test` method in `AnsiCastSuiteBase` to automatically add the "ANSI mode:" prefix in the test name.

**Review by cloud-fan — APPROVED** [review]
+1!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AnsiCastSuiteBase.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AnsiCastSuiteBase.scala:54]: Hmm, that will affect the common test cases in CastSuiteBase which are not ANSI-mode-specific behavior.

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/AnsiCastSuiteBase.scala:54]: oh I see

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by gengliangwang** [comment]
Merging to master

**Comment by MaxGekk** [comment]
@gengliangwang I have huge conflicts in my PR https://github.com/apache/spark/pull/32909 after rebasing on this. Just to double check, did you just move the code around w/o any modifications?

**Comment by gengliangwang** [comment]
@MaxGekk Sorry about that. Yes I just moved code around. I checked #32909  and I think you can consider removing/stashing the changes in `CastSuite.scala` and apply them after rebasing.

**Pull request closed without merging**